pynapple/process/signal_processing.py

```diff
@@ -0,0 +1,268 @@
+            t=sig.index, d=cwt.reshape(output_shape), time_support=sig.time_support
+        )
+
+    return nap.TsdTensor(
+        t=sig.index, d=cwt.reshape(output_shape), time_support=sig.time_support
+    )
+
+
+def generate_morlet_filterbank(freqs, fs, n_cycles=1.5, scaling=1.0, precision=10):
+    """
+    Generates a Morlet filterbank using the given frequencies and parameters. Can be used purely for visualization,
+    or to convolve with a pynapple Tsd, TsdFrame, or TsdTensor as part of a wavelet decomposition process.
+
+    Parameters
+    ----------
+    freqs : 1d array or list of float
+        If array, frequency values to estimate with morlet wavelets.
+        If list, define the frequency range, as [freq_start, freq_stop, freq_step].
+        The `freq_step` is optional, and defaults to 1. Range is inclusive of `freq_stop` value.
+    fs : float
+        Sampling rate, in Hz.
+    n_cycles : float or 1d array
+        Length of the filter, as the number of cycles for each frequency.
+        If 1d array, this defines n_cycles for each frequency.
+    scaling : float
+        Scaling factor.
+    precision: int.
+        Precision of wavelet to use.
+
+    Returns
+    -------
+    filter_bank : pynapple.TsdFrame
+        list of Morlet wavelet filters of the frequencies given
+    """
+    if len(freqs) == 0:
+        raise ValueError("Given list of freqs cannot be empty.")
+    filter_bank = []
+    time_cutoff = 8
+    morlet_f = _morlet(int(2**precision), ncycles=n_cycles, scaling=scaling)
+    x = np.linspace(-time_cutoff, time_cutoff, int(2**precision))
+    int_psi = np.conj(_integrate(morlet_f, x[1] - x[0]))
+    max_len = -1
+    for freq in freqs:
+        scale = scaling / (freq / fs)
+        j = np.arange(scale * (x[-1] - x[0]) + 1) / (scale * (x[1] - x[0]))
+        j = j.astype(int)  # floor
+        if j[-1] >= int_psi.size:
+            j = np.extract(j < int_psi.size, j)
+        int_psi_scale = int_psi[j][::-1]
+        if len(int_psi_scale) > max_len:
+            max_len = len(int_psi_scale)
+            time = np.linspace(
+                -time_cutoff * scaling / freq, time_cutoff * scaling / freq, max_len
+            )
+        filter_bank.append(int_psi_scale)
+    filter_bank = [
+        np.pad(
+            arr,
+            ((max_len - len(arr)) // 2, (max_len - len(arr) + 1) // 2),
+            constant_values=0.0,
+        )
+        for arr in filter_bank
+    ]
+    return nap.TsdFrame(d=np.array(filter_bank).transpose(), t=time)
+
+
+def _integrate(arr, step):
+    """
+    Integrates an array with respect to some step param. Used for integrating complex wavelets.
+
+    Parameters
+    ----------
+    arr : np.ndarray
+        wave function to be integrated
+    step : float
+        Step size of vgiven wave function array
+
+    Returns
+    -------
+    array
+        Complex-valued integrated wavelet
+
+    """
+    integral = np.cumsum(arr)
+    integral *= step
```

Comment:
I don't know how much precision matters here, but you can use the simpson method to get a better approximation of the numerical integral. may be slow to do it in a cumulative fashion but something to consider


```python
from scipy.integrate import simpson
import numpy as np
a,b = 0,1
func = lambda x: np.sin(x)
integr = lambda x: -np.cos(x)

xs = np.linspace(a,b,20)
print(f"true integral: {integr(b) - integr(a)}")
print(f"simps integral: {simpson(func(xs), x=xs)}")
print(f"euler integral: {np.sum(func(xs))*(xs[1] - xs[0])}")

# true integral: 0.45969769413186023
# simps integral: 0.4596975270737199
# euler integral: 0.48173554578218836
```

Reply:
The precision parameter controls the amount of time points at which the Morlet wavelet is evaluated at. One Morlet wavelet is created with 2**precision time points, and then it is stretched to calculate wavelet coefficients at different frequencies. This is why, at low frequencies (where the wavelet must be stretched a lot), with a low frequency you can see the wavelet repeating values.